design/design.md

```diff
@@ -0,0 +1,232 @@
+# Design for Streamly Coreutils package
+
+* cp
+
+  ```
+      data CpOptions = CpOptions {
+         attributesOnly :: Bool,
+         interactive :: Bool,
+         parents :: Bool,
+         symbolicLink :: Bool,
+         noTargetDir :: Bool,
+         verbose :: Bool,
+         noTargetDir :: Bool
+         ...
+      }
+
+      cpFile :: CpOptions -> SomeBase File -> SomeBase Dir -> SomeBase File -> IO ()
+
+  ```
+
+  1. Similar to srcFile -> destDir -> destFile
+  2. Copies the source file to a dest directory with dest file name
+  3. If dest file name is empty, then it becomes File -> Dir case (with same file name as source)
+
+  4. For File -> File within same directory, current directory should be passed as destDir argument
+   Else we can also say that an empty Dest Dir argument is equivalent to copying in the same directory
+
+  ```
+      cpFiletoDir :: (IsStream t, Monad m) => CpOptions -> t m (SomeBase File) -> SomeBase Dir -> IO ()
+         uses cpFile
+         (stream of input files so that copying can be done concurrently with use of appropriate streams)
+         or map can be used as well
+
+      cpDirToDir :: (IsStream t, Monad m) => CpOptions -> SomeBase Dir -> SomeBase Dir -> IO ()
+         should create a stream of files (t m (SomeBase File)) and call cpFileToDir
+
```

Comment:
Here are some suggestions for `cp`:

`mv` should also be made

`cpFile :: Path File -> Path Dir` -- Cannot be an empty Dir

`cpFiles :: Stream (Path File) -> Path Dir` -- Can be made concurrent

`cpDir :: Path Dir -> Path Dir`

`cpFileWithMv :: Path File -> Path File` -- combine `cpFile` and `mv`

Reply:
Why is there a need for `cpFileWithMv` ?